### PR TITLE
Improve query validation robustness

### PR DESCRIPTION
**Pull request by leoyvens**
A few changes to improve guarding against overflows:
- Explicitly validate against cyclical fragments.
- Validate complexity before fields.
- Validate query depth even if the complexity is unlimited.
- Add a test that previously caused a stack overflow.

### CODE REVIEW

File: core/tests/interfaces.rs [hunk 1]
```diff
@@ -1250,3 +1250,63 @@ fn enums() {
...
+            ...FooFrag
+            }
+        }
+
+        fragment FooFrag on Foo {
+        id
+        bar {
+            ...BarFrag
+        }
+        }
```
Comment (Jannis) [inline core/tests/interfaces.rs:1307]: The indentation here is a little off.

**Review by Jannis — APPROVED** [review]
Looks good!
